### DISCUSSION TITLE
Concept hierarchy for OMOP Conditions

### DISCUSSION THREAD

**Discussion by lrasmus** [uwrit/leaf#438] [Q&A]
Thanks so much for the excellent [leaf-scripts for OMOP](https://github.com/uwrit/leaf-scripts/tree/master/omop/5.3).  I am sorry to ask for 'more, more, more', but before working on this myself I was wondering if anyone had created a Leaf Concept hierarchy for OMOP Conditions?  Given that OMOP relies on SNOMED for conditions, I was curious in part of any SNOMED-based hierarchy for conditions is the best way to go, or if using an ICD-based hierarchy is better.

Happy to contribute back anything we create if a solution doesn't already exist.

**Comment by ndobb** [comment]
Hi @lrasmus,

I'm glad those scripts have been useful and that's totally fine. @artgoldberg and I and others have discussed strategies for diagnoses a bit and I think many have questions similar to this.

My personal opinion is that users are >90% of the time going to be much more comfortable using and interpreting ICD{9|10} codes rather than SNOMED, given widespread use in EHRs and billing, etc. Given that there there are also far less ICD-10 codes than SNOMED, for example, it's also more compact and presumably easier to navigate than a SNOMED hierarchy, potentially.

Given those assumptions, we haven't found a nice one-size-fits-all solution as different sites have different preferences and may populate their OMOP DBs differently, for example using SNOMED vs. ICD-10 for concept mapping, and populating the `source_condition_value` with ICD-10 or not, etc. So what the `WHERE` clause does in Leaf potentially can vary quite a bit. Last, if I recall correctly when last I checked the ICD-10 tree was not fully represented in the OMOP `concept_ancestor` table, so we weren't able to properly leverage that as we'd like.

@lrasmus in your OMOP DB, how are you populating conditions?

**Reply by artgoldberg** [reply]
Hi @ndobb and @lrasmus 

@lrasmus I'm currently working to create a diagnosis (condition) hierarchy for Leaf for our omop CDM at [Mount Sinai](https://www.mountsinai.org/). As Nic suggests, I'm making a hierarchy of ICD-10-CM codes for users. That's the code they best understand. Since we have other standard OHDSI applications accessing the same CDM, we must put standard OMOP SNOMED codes in `condition_occurrence.condition_concept_id`. And since we're tracking source code provenance, we'll put Epic diagnosis information in the `condition_source_concept_*` fields in `condition_occurrence`.

To make the queries work we need to:

1. Map 'Epic diagnosis ID' → ICD-10-CM, which we do by taking the 1-to-1 mappings from our Epic database and spot checking them manually.
2. Use [Athena's](https://athena.ohdsi.org/search-terms/start) reference data to map ICD-10-CM → SNOMED
3. Load the ICD10 → SNOMED mappings into `concept_relationship`
4. Load the SNOMED codes indicated by the Epic → ICD10 → SNOMED into `condition_occurrence.condition_concept_id`

Given this, our Leaf queries will need to access the `condition_occurrence`, `concept_relationship`, and `concept` tables.

More details in my next comment.

Arthur

**Reply by artgoldberg** [reply]
Hi @ndobb

Let's dive into how Leaf diagnosis concepts will query our CDM. I propose this design, which is based on an actual Leaf query that involves one concept, and raise a few questions. This example shows a search for a diagnosis represented by an internal node in the ICD10CM hierarchy.

    SELECT
        P0.person_id
    FROM
        (
            SELECT
                COND_O.person_id

            FROM
                concept AS CONCEPT,
                concept_relationship AS CONCEPT_REL,
                condition_occurrence AS COND_O

            WHERE
                /* Atypical virus infections of central nervous system (ICD10CM:A81.00-A81.9) */
                CONCEPT.concept_code BETWEEN 'A81.00' AND 'A81.9' AND
                -- the SNOMED code in COND_O.condition_concept_id must map to
                -- the selected ICD10CM code(s) in CONCEPT
                CONCEPT.vocabulary_id = 'ICD10CM' AND
                CONCEPT.concept_id = CONCEPT_REL.concept_id_2 AND
                CONCEPT_REL.relationship_id = 'Maps to' AND
                CONCEPT_REL.concept_id_1 = COND_O.condition_concept_id

            GROUP BY
                COND_O.person_id

        ) AS P0

The only part of this query that differs between diagnosis concepts are the lines

                /* Atypical virus infections of central nervous system (ICD10CM:A81.00-A81.9) */
                CONCEPT.concept_code BETWEEN 'A81.00' AND 'A81.9' AND

First question: I hope that this query be represented in the LeafDB. Roughly, the associated `ConceptSqlSet` must contain `condition_occurrence` in `SqlSetFrom`, and `Concept.SqlSetWhere` must contain the entire `WHERE` clause. My question: Where should the rest of the `FROM` clause go?

Second question: Should the query add conditions to ensure that `CONCEPT_REL.concept_id_1` is a SNOMED code? (Alternatively, `concept_relationship` could be constrained to map only SNOMED codes to ICD10CM but I reject this option because one of our principles is that Leaf should not and cannot modify the CDM since many other tools access it.) I think the answer should be _yes_, because otherwise the query could be wrong if codes from vocabularies other than SNOMED map to ICD10CM in `concept_relationship`. If it is _yes_, then the   `FROM` clause also needs `concept AS CONCEPT_SNOMED`, and the `WHERE` must be extended with:

            AND
            CONCEPT_REL.concept_id_1 = CONCEPT_SNOMED.concept_id AND
            CONCEPT_SNOMED.vocabulary_id = 'SNOMED'

This portion of the query is also identical for all diagnosis concepts.

Third question: Since Leaf users query MSDW2 interactively they require good performance. How should the CDM be indexed to ensure that this query performs well? Given that all Leaf diagnosis queries will have this structure and that the CDM is modified only during nightly incremental updates, in my view the CDM tables should include these indices:

Column(s) indexed | Index type | Rationale
------------ | ------------- | -------------
`concept.concept_code` | Hash | Obtain O(1) access a concept row when queried with = operator
`concept.concept_code` | Clustered | Quickly access to concept rows when queried with BETWEEN operator
`concept_relationship.concept_id_2` | Hash | Obtain O(1) join between this column and concept.concept_id
`condition_occurrence.condition_concept_id` | Hash | Obtain O(1) join between concept_relationship.concept_id_1 and this column
`concept.concept_id` | Hash | Obtain O(1) join between concept_relationship.concept_id_1 and this column

These indices assume 1) a Clustered index does not exist on `concept`, 2) that concept queries are transformed into query plans that first apply the `CONCEPT.concept_code` condition (which I expect the plan generator will find most restrictive) and then add other fields and tables by using rows that have been found or applying join conditions.

Regards
Arthur

PS. Happy to move this to another discussion if it's too much here.

**Reply by ndobb** [reply]
This is the perfect place to discuss, @artgoldberg. I'd suggest we structure queries like the following (omitting database name and schema for simplicity):

```sql
SELECT 
       CO.person_id
FROM 
       condition_occurrence AS CO  /* <-- SqlSetFrom */
WHERE 

       /* Begin SqlSetWhere */ 

       EXISTS                      
	   (SELECT 
                 1
	    FROM 
                 concept AS C_SNOMED
		 INNER JOIN concept_relationship AS CR
                      ON C_SNOMED.concept_id = CR.concept_id_2
		 INNER JOIN concept AS C_ICD10
		      ON C_ICD10.concept_id = CR.concept_id_1
	    WHERE 
                 C_SNOMED.vocabulary_id = 'SNOMED'
		 AND CR.relationship_id = 'Maps to'
		 AND C_ICD10.vocabulary_id = 'ICD10CM'
	         AND C_ICD10.concept_code BETWEEN 'A81.00' AND 'A81.9' /* <-- depends on Concept */
		 AND CO.condition_concept_id = C_SNOMED.concept_id)

       /* End SqlSetWhere */ 

GROUP BY 
       CO.person_id
```

Configuring with the Leaf alias symbol, `"@"`, the `SqlSetWhere` in the Leaf Concept would thus be:

```sql
EXISTS                      
	   (SELECT 
                 1
	    FROM 
                 concept AS @C_SNOMED
		 INNER JOIN concept_relationship AS @CR
                      ON @C_SNOMED.concept_id = @CR.concept_id_2
		 INNER JOIN concept AS @C_ICD10
		      ON @C_ICD10.concept_id = @CR.concept_id_1
	    WHERE 
                 @C_SNOMED.vocabulary_id = 'SNOMED'
		 AND @CR.relationship_id = 'Maps to'
		 AND @C_ICD10.vocabulary_id = 'ICD10CM'
	         AND @C_ICD10.concept_code BETWEEN 'A81.00' AND 'A81.9'
		 AND @.condition_concept_id = @C_SNOMED.concept_id)
```

Results using this as a proof-of-concept in a test Leaf instance on our EDW suggests this is a reasonable approach:
![image](https://user-images.githubusercontent.com/10424962/127197338-c660ac8f-081a-4b21-a4a6-2565a2706c49.png)

Regarding your questions:
1. *Where should the rest of the FROM clause go?* 
    
    **Using the above example as a guide, we can put all of the logic in the WHERE clause in this way (and the `FROM` simply be `condition_occurrence`). Note that we could likely go even further with nested EXISTS statements to squeeze out even better performance, though it becomes a bit harder to make sense of the query:**

![image](https://user-images.githubusercontent.com/10424962/127198760-47856a2b-75ef-42dc-b8cd-7c6e69617525.png)

2. *Should the query add conditions to ensure that `CONCEPT_REL.concept_id_1` is a SNOMED code?*
    
    **Yes, I agree and think we should, for the reasons you state. If for some reason another site wanted to use this approach but using a coding system other than SNOMED, it would be straightforward for them to adjust accordingly.**

3. *How should the CDM be indexed to ensure that this query performs well?*
    
    **This is a really good question. I need to experiment a bit more but so far it seems that queries following the above approach perform reasonably well using the "out-of-the-box" OMOP indices against our EDW. I'll do so and respond again. I think the proposed indexed columns and types make sense.**

**Reply by artgoldberg** [reply]
Your response is great @ndobb . As I said to my wife, Leaf has triangular holes and Nic explained how to convert my round peg query into a triangular one that fits into Leaf. Very nice.

My next steps are to obtain local access to the data that maps 'Epic diagnosis ID' → ICD-10-CM, revise the Leaf diagnosis script to generate the code you outlined above, and then use the data and Athena's mappings to drive the script.

Arthur

**Reply by ndobb** [reply]
Sounds great! Let us know how things go and if anything doesn't work as expected, @artgoldberg.

**Reply by lrasmus** [reply]
Thanks both @ndobb and @artgoldberg, this is really helpful information!  Our setup is going to be very similar to what @artgoldberg is doing - we are handling the ICD9/10 -> SNOMED mapping for conditions within our OMOP ETL, and so we can replicate the ICD10 hierarchy approach for people to query from.

**Reply by ndobb** [reply]
Awesome! Like Arthur please let me know how things go, @Irasmus.

**Reply by artgoldberg** [reply]
Hi Folks

This is my high level plan:
1. Fill in the `condition_occurrence.condition_concept_id` fields and `concept_relationship` records needed in our CDM by the script that will load the LeafDB with the data needed by the Leaf conditions queries described above. This involves creating EPIC EDG .1 → ICD10 → SNOMED mappings. Roughly, the first relationship is taken from Caboodle and the second is taken from Athena.
2. Revise, debug and run the SQL script that will load the LeafDB.
3. Test Leaf.

@ndobb , I have completed step 1, and find 49,503 unique ICD10 codes. Would that be a reasonable number of codes to deploy in the Leaf condition concept hierarchy? I see that the `UMLS_ICD10` table contains 95,671 rows, so I'm glad the two have the same order of magnitude.

Arthur

**Reply by ndobb** [reply]
@artgoldberg, I just checked our prod UW Leaf instance and UMLS DB (which is up to date, to my knowledge). I'm seeing 72,616 unique ICD10 codes in UMLS and slightly less in our Leaf instance (which probably is overdue for adding newer ICD10 codes). The `UMLS_ICD10` table includes *ranges* of codes in a hierarchical format, so it makes sense that that would have more rows.

So I would say 72,616 ICD10 codes should be roughly what we should aim for, using UMLS as our gold standard. Any idea what codes could be being dropped? I'm happy to help debug if feasible.

For reference, the UMLS query I'm doing is:

```sql
SELECT COUNT(DISTINCT CODE)
FROM UMLS.MRCONSO
WHERE SAB = 'ICD10CM'
      AND TTY = 'PT' /* Preferred Term */
```

**Reply by lrasmus** [reply]
One consideration I've wrestled with before, do we include all ICD10 codes (and just let people know if there are 0 patients with any code), or do we only load the ICD10 codes that have been assigned to patients?  I've opted for the former, just because I can imagine confusion searching for an ICD10 code to perform a query and the code doesn't exist.

I mention it wondering if @artgoldberg is seeing fewer codes because if the source isn't the UMLS, if it's only including a subset of ICD10 codes that could be assigned (meaning it may have already filtered out codes that aren't allowed for billing)?

**Reply by ndobb** [reply]
Ah that's a really good point, @lrasmus. At UW we include even codes that have never been used (ie, zero patients) so our users can distinguish between codes that haven't been given versus codes that haven't yet been added to Leaf.

**Reply by artgoldberg** [reply]
You've hit the nail on the head @lrasmus . My more clinical colleagues have recommended that we filter out diagnoses that have not been assigned to patients in the EHR. My 49,503 unique ICD10 codes includes that filter.

But if I obtain all ICD10 codes, including ones that map to Epic codes that are have not been used by clinicians, I find 95,506 unique ICD10 codes in our clinical DW. Given your point @ndobb , I'll revisit that recommendation by my colleagues.

Overall, I'm comfortable with the number of unique ICD10 codes we have. They're in the neighborhood of what you find Nic, and our number of active codes makes sense.

Arthur

**Reply by artgoldberg** [reply]
Hello @ndobb and @lrasmus 

I've been thinking about this discussion. We should more carefully analyze the decision whether to include ICD10CM codes that map to Epic codes that have not been used by clinicians.

Not including these ICD10CM codes only changes the CDM's `concept_relationship` table. Leaving them out does not change the `LeafDB.app.Concept` table, because it is determined by the `UMLS_ICD10` table, which depends only on the UMLS table [`MRHIER.RRF`](https://www.ncbi.nlm.nih.gov/books/NBK9685/table/ch03.T.computable_hierarchies_file_mrhie/) that maps ICD10CM concepts into a hierarchy. Thus, the structure of the Leaf interface depends only on `MRHIER.RRF`.

Since a complete `concept_relationship` table may help other software that uses the CDM, I conclude that all 1-to-1 mappings between Epic EDG .1 codes and ICD10CM codes should be included in it.

Arthur
